server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -222,48 +229,80 @@ protected void runTasks(TaskInputs taskInputs) {
 
                 logger.debug("publishing cluster state version [{}]", newClusterState.version());
-                try {
-                    clusterStatePublisher.accept(clusterChangedEvent, taskOutputs.createAckListener(threadPool, newClusterState));
-                } catch (Discovery.FailedToCommitClusterStateException t) {
-                    final long version = newClusterState.version();
-                    logger.warn(() -> new ParameterizedMessage(
-                            "failing [{}]: failed to commit cluster state version [{}]", summary, version), t);
-                    taskOutputs.publishingFailed(t);
-                    return;
-                }
-
-                taskOutputs.processedDifferentClusterState(previousClusterState, newClusterState);
-
-                try {
-                    taskOutputs.clusterStatePublished(clusterChangedEvent);
-                } catch (Exception e) {
-                    logger.error(() -> new ParameterizedMessage(
-                            "exception thrown while notifying executor of new cluster state publication [{}]",
-                            summary), e);
-                }
-                TimeValue executionTime = TimeValue.timeValueMillis(Math.max(0, TimeValue.nsecToMSec(currentTimeInNanos() - startTimeNS)));
-                logger.debug("processing [{}]: took [{}] done publishing updated cluster state (version: {}, uuid: {})", summary,
-                    executionTime, newClusterState.version(),
-                    newClusterState.stateUUID());
-                warnAboutSlowTaskIfNeeded(executionTime, summary);
+                publish(clusterChangedEvent, taskOutputs, startTimeNS);
             } catch (Exception e) {
-                TimeValue executionTime = TimeValue.timeValueMillis(Math.max(0, TimeValue.nsecToMSec(currentTimeInNanos() - startTimeNS)));
-                final long version = newClusterState.version();
-                final String stateUUID = newClusterState.stateUUID();
-                final String fullState = newClusterState.toString();
-                logger.warn(() -> new ParameterizedMessage(
-                        "failed to publish updated cluster state in [{}]:\nversion [{}], uuid [{}], source [{}]\n{}",
-                        executionTime,
-                        version,
-                        stateUUID,
-                        summary,
-                        fullState),
-                    e);
-                // TODO: do we want to call updateTask.onFailure here?
+                handleException(summary, startTimeNS, newClusterState, e);
             }
         }
     }
 
+    protected void publish(ClusterChangedEvent clusterChangedEvent, TaskOutputs taskOutputs, long startTimeNS) {
+        final PlainActionFuture<Void> fut = new PlainActionFuture<Void>() {
+            @Override
+            protected boolean blockingAllowed() {
+                // allow this one to block on the MasterServiceUpdateThread
```

Comment:
This comment would be unnecessary if we wrote something like:

    return Thread.currentThread().getName().contains(MASTER_UPDATE_THREAD_NAME) || super.blockingAllowed();